interfaces/builtin/zigbee_dongle.go

```diff
@@ -0,0 +1,137 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+// ZigbeeDongleInterface is the type for serial port interfaces.
+type ZigbeeDongleInterface struct{}
+
+// Name of the zigbee-dongle interface.
+func (iface *ZigbeeDongleInterface) Name() string {
+	return "zigbee-dongle"
+}
+
+func (iface *ZigbeeDongleInterface) String() string {
+	return iface.Name()
+}
+
+var zigbeeDongleDevPath = "/dev/zigbee/*"
+
+var zigbeeDonglePermanentSlotUdev = []byte(`
+IMPORT{builtin}="usb_id"
+SUBSYSTEM=="tty", SUBSYSTEMS=="usb", ATTRS{idProduct}=="0003", ATTRS{idVendor}=="10c4", SYMLINK+="zigbee/$env{ID_SERIAL}"
+`)
```

Comment:
So, this might be the time to use the Udev backend for one of its initial design goals: adding devices to an app-specific device cgroup. This would address the concerns of @zyga and allow fine-grained control. Here is how it could work:
1. this interface has attributes for idProduct and idVendor (feel free to name the yaml keys something different of course)
2. when attributes are used, fill in ATTRS{idProduct} and ATTRS{idVendor} appropriately, add a TAG+="snap_connecting_app" for each connected plug. udev TAGs cannot use globs so you'll need a rule for each. The TAG to add is simply the security label for the plug replacing '.' with '_'. Use /dev/** for apparmor.
3. if it makes sense, when attributes are not used, don't add TAGs and simply add all known idProducts and idVendors for zigbee devices to zigbeeDonglePermanentSlotUdev and use /dev/zigbee/* for apparmor

What does this do and why does it work? When attributes are not used, then access is granted to all known zigbee devices and this access is enforced with the /dev/zigbee/* apparmor glob rule. When attributes are used, access is granted only to those devices matching the idProduct and idVendor and this is enforced by adding only those devices to the snap's device cgroup (we use a /dev/** apparmor rule since we rely on the device cgroup for security instead of apparmor). In both cases hot plugging and unplugging should work as expected.

A consideration with this approach is that using this interface with other interfaces that grant access to /dev/* via apparmor means that those other devices won't be in the device cgroup. For example, suppose we have the following yaml:
```
plugs:
  - camera
  - zigbee
    id-product: "0003"
    id-vendor: "10c4"
```
With the above we end up with apparmor policy (which while /dev/video0 is redundant, is still valid):
  /dev/video0 rw,
  /dev/** rw,

but only the zigbee devices are added to the device cgroup so while apparmor allows access to /dev/video0, the snap won't be able to access /dev/video0 because /dev/video0 wasn't udev tagged and therefore snap-confine didn't add it to the launcher. Importantly, this issue will not be unique to the zigbee interface-- it will need to be fixed to support TAGging for device cgroups which I think will be particularly important for gadget snap labelling, etc.

OTOH to fix this we would have to abstract out the /dev/ accesses that are sprinkled through the interfaces and for each encode both an apparmor rule and a udev snippet in the interfaces, then when generating the security policy detect when a device cgroup will be used and add the udev snippets when it is and the apparmor rules when not.

Since the default with no attributes is this PR, perhaps it would make sense to commit this as is (ie, just the /dev/zigbee/* apparmor rule) and then work through attributes and specific device assignment in a future PR. As such, I agree this should not be renamed zigbee-dongles.